ouroboros-network/src/Ouroboros/Network/Connections/Socket/Types.hs

```diff
@@ -0,0 +1,141 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE RankNTypes #-}
+
+module Ouroboros.Network.Connections.Socket.Types
+  ( SockType (..)
+  , SockAddr (..)
+  , ConnectionId (..)
+  , makeConnectionId
+  , connectionIdPair
+  , Some (..)
+  , someSockType
+  , withSockType
+  , matchSockType
+  , matchSockAddr
+  , forgetSockType
+  ) where
+
+import Data.Kind (Type)
+import qualified Network.Socket as Socket
+
+data SockType where
+  IPv4 :: SockType
+  IPv6 :: SockType
+  Unix :: SockType
+
+-- | Like `Network.Socket.SockAddr` but with a type parameter indicating which
+-- kind of address it is: IPv4, IPv6, or Unix domain.
+data SockAddr (sockType :: SockType) where
+  SockAddrIPv4 :: !Socket.PortNumber -> !Socket.HostAddress -> SockAddr IPv4
+  SockAddrIPv6 :: !Socket.PortNumber
+               -> !Socket.FlowInfo
+               -> !Socket.HostAddress6
+               -> !Socket.ScopeID
+               -> SockAddr IPv6
+  -- No strictness here because the Network.Socket analogue is not strict.
+  SockAddrUnix :: String -> SockAddr Unix
+
+instance Eq (SockAddr sockType) where
+  x == y = forgetSockType x == forgetSockType y
+
+instance Show (SockAddr sockType) where
+  show = show . forgetSockType
+
+-- | A connection is identified by a pair of addresses. For IPv* this is fine,
+-- but not for Unix domain sockets: these can be unnamed, so that a connecting
+-- socket which does not bind will get the address `SockAddrUnix ""`.
+--
+-- How to deal with this?
+-- One obvious way is to simply make up an ephemeral identifier for any
+-- `SockAddrUnix ""` term, maybe using its file descriptor.
+--
+data ConnectionId where
+  ConnectionIdIPv6 :: !(SockAddr IPv6) -> !(SockAddr IPv6) -> ConnectionId
+  ConnectionIdIPv4 :: !(SockAddr IPv4) -> !(SockAddr IPv4) -> ConnectionId
+  ConnectionIdUnix :: !(SockAddr Unix) -> !(SockAddr Unix) -> ConnectionId
+
+deriving instance Show ConnectionId
+
+instance Eq ConnectionId where
```

Comment:
This and the `Ord` instance are actually wrong I believe. The order of the tuple should not matter. A connection from `hostA:portA` to `hostB:portB` is also a connection from `hostB:portB` to `hostA:portA`. Could use the `Ord SockAddr => Ord (Set SockAddr)` instance to get this:

```
Set.fromList [forgetSockType a, forgetSockType b] == Set.fromList [forgetSockType c, fogetSockType d]
Set.fromList [forgetSockType a, forgetSockType b] `compare` Set.fromList [forgetSockType c, fogetSockType d]
```

Reply:
I think that the order of the tuples should have meaning, or rather the socket fields in the Connection should have meaning. One field should correspond to the local address and one field should correspond to the remote address so that one doesn't have to guess when interpreting logs in the future.

Reply:
Right. Maybe the better solution is to always choose local as the first component

Reply:
I suggest we use record fields then which one can use with GADTs.  It will make it clear which one is the local and which one is the remote.